File: backend/app/schemas/network.py
from datetime import datetime

from pydantic import field_serializer, field_validator
from sqlmodel import SQLModel

from app.utils import validate_optional_http_url, validate_required_http_url

from app.schemas.coordinates import Latitude, Longitude


class NodeStats(SQLModel):
    users: int = 0
    projects: int = 0
    collections: int = 0
    audios: int = 0
    photos: int = 0
    videos: int = 0
    annotations: int = 0
    sites: int = 0


class NetworkNodePublic(SQLModel):
    id: int
    name: str
    app_url: str
    latitude: float | None
    longitude: float | None
    is_local: bool
    shared: bool
    stats: NodeStats
    last_synced_at: datetime | None

    @field_serializer("last_synced_at")
    def serialize_last_synced_at(self, dt: datetime | None) -> str | None:
        if dt is None:
            return None
        return dt.strftime("%Y-%m-%d %H:%M:%S")


class NetworkSettings(SQLModel):
    server_name: str = ""
    app_url: str = ""
    host_url: str = ""
    latitude: float | None = None
    longitude: float | None = None
    shared: bool = False
    federation_secret: str = ""


class NetworkSettingsUpdate(SQLModel):
    server_name: str | None = None
    app_url: str | None = None
    host_url: str | None = None
    latitude: Latitude | None = None
    longitude: Longitude | None = None
    shared: bool | None = None
    federation_secret: str | None = None

    @field_validator("app_url", "host_url")
    @classmethod
    def validate_urls(cls, value: str | None) -> str | None:
        return validate_optional_http_url(value)


class NodeRegistration(SQLModel):
    app_url: str
    name: str
    latitude: Latitude | None = None
    longitude: Longitude | None = None
    stats: NodeStats | None = None
    shared: bool = True

    @field_validator("app_url")
    @classmethod
    def validate_app_url(cls, value: str) -> str:
        return validate_required_http_url(value)


class SyncResult(SQLModel):
    synced: int
    message: str = "ok"

File: frontend/src/api/endpoints/network.ts
import { apiClient } from "../client"

/** GET /v1/network-nodes — `stats` 与后端 NodeStats 一致 */
export interface NodeStats {
    users: number
    projects: number
    collections: number
    audios: number
    photos: number
    videos: number
    annotations: number
    sites: number
}

/**
 * GET /v1/network-nodes 列表项。
 * 本地实例节点可能使用 `id: 0`；请以 `is_local` / `id` 判断，勿用 `if (id)` 这类会把 0 判假的写法。
 */
export interface NetworkNodePublic {
    id: number
    name: string
    app_url: string
    latitude: number | null
    longitude: number | null
    /** 当前实例目录节点为 true（首页地图橙色）；联邦种子节点为 false（绿色） */
    is_local: boolean
    stats: NodeStats
    last_synced_at: string | null
}

/** GET /v1/network-settings — 后端为管理员接口；无权限时请求会失败，前端需降级 */
export interface NetworkSettings {
    server_name: string
    app_url: string
    host_url: string
    latitude: number | null
    longitude: number | null
    shared: boolean
    federation_secret: string
}

/** PUT /v1/network-settings — 字段均为可选，与后端 NetworkSettingsUpdate 一致 */
export interface NetworkSettingsUpdate {
    server_name?: string
    app_url?: string
    host_url?: string
    latitude?: number | null
    longitude?: number | null
    shared?: boolean
    federation_secret?: string | null
}

export const networkApi = {
    /** 获取网络节点列表（地图标记 + 统计） */
    getNodes() {
        return apiClient.get<{ code: number; message: string; data: NetworkNodePublic[] }>("/v1/network-nodes")
    },

    /** 获取本实例联邦配置（含地图锚点经纬度）；需管理员，公开首页调用失败属正常 */
    getSettings() {
        return apiClient.get<{ code: number; message: string; data: NetworkSettings }>("/v1/network-settings")
    },

    /** 更新联邦配置；需管理员 */
    updateSettings(body: NetworkSettingsUpdate) {
        return apiClient.put<{ code: number; message: string; data: NetworkSettings }>("/v1/network-settings", body)
    },

    /** 生成新的联邦密钥并保存；需管理员 */
    generateFederationSecret() {
        return apiClient.post<{ code: number; message: string; data: NetworkSettings }>(
            "/v1/network-secret-rotations",
        )
    },
}
